robosuite/utils/video_utils.py

```diff
@@ -0,0 +1,316 @@
+
+    def grab_next_frame(self):
+        """
+        Grabs the next frame in the demo sequence by stepping the simulation and returning the resulting value(s)
+
+        Returns:
+            dict: Keyword-mapped np.arrays from the demonstration sequence, corresponding to all image modalities used
+                in the playback environment (e.g.: "image", "depth", "segmentation_instance")
+        """
+        # Make sure the episode isn't completed yet, if so, we load the next episode
+        if self.step == self.n_steps:
+            self.load_episode_xml(demo_num=self.current_ep + 1)
+
+        # Step the environment and grab obs
+        if self.replay_from_actions:
+            obs, _, _, _ = self.env.step(self.actions[self.step])
+        else:           # replay from states
+            self.env.sim.set_state_from_flattened(self.states[self.step + 1])
+            if self.use_dr:
+                self.env.step_randomization()
+            self.env.sim.forward()
+            obs = self.env._get_observation()
+
+        # Increment the step counter
+        self.step += 1
+
+        # Return all relevant frames
+        return {k.split(f"{self.camera}_")[-1]: obs[k] for k in obs if self.camera in k}
+
+    def grab_episode_frames(self, demo_num, pan_point=(0, 0, 0.8), pan_axis=(0, 0, 1), pan_rate=0.01):
+        """
+        Playback entire episode @demo_num, while optionally rotating the camera about point @pan_point and
+            axis @pan_axis if @pan_rate > 0
+
+        Args:
+            demo_num (int): Demonstration episode number to load for playback
+            pan_point (3-array): (x,y,z) cartesian coordinates about which to rotate camera in camera frame
+            pan_direction (3-array): (ax,ay,az) axis about which to rotate camera in camera frame
+            pan_rate (float): how quickly to pan camera if not 0
+
+        Returns:
+            dict: Keyword-mapped stacked np.arrays from the demonstration sequence, corresponding to all image
+                modalities used in the playback environment (e.g.: "image", "depth", "segmentation_instance")
+
+        """
+        # First, load env
+        self.load_episode_xml(demo_num=demo_num)
+
+        # Initialize dict to return
+        obs = self.env._get_observation()
+        frames_dict = {k.split(f"{self.camera}_")[-1]: [] for k in obs if self.camera in k}
+
+        # Continue to loop playback steps while there are still frames left in the episode
+        while self.step < self.n_steps:
+            # Take playback step and add frames
+            for k, frame in self.grab_next_frame().items():
+                frames_dict[k].append(frame)
+
+            # Update camera pose
+            self.rotate_camera(point=pan_point, axis=pan_axis, angle=pan_rate)
+
+        # Stack all frames and return
+        return {k: np.stack(frames) for k, frames in frames_dict.items()}
+
+    def modify_xml_for_camera_movement(self, xml, camera_name):
```

Comment:
these seem to be copied from the camera movement script. We should probably avoid duplication, and change that script to use this function instead

Reply:
Agreed. What exact script are you referring to? I don't recall adding a script with duplicate functionality